crates/ruff_python_parser/src/parser/statement.rs

```diff
@@ -2125,48 +2134,86 @@ impl<'src> Parser<'src> {
         while self.at(TokenKind::At) {
             progress.assert_progressing(self);
+
             let decorator_start = self.node_start();
             self.bump(TokenKind::At);
 
+            // test_err decorator_invalid_expression
+            // @*x
+            // @(*x)
+            // @((*x))
+            // @yield x
+            // @yield from x
+            // def foo(): ...
             let parsed_expr = self.parse_named_expression_or_higher(AllowStarredExpression::No);
+
             decorators.push(ast::Decorator {
                 expression: parsed_expr.expr,
                 range: self.node_range(decorator_start),
             });
 
+            // test_err decorator_missing_newline
+            // @x def foo(): ...
+            // @x async def foo(): ...
+            // @x class Foo: ...
             self.expect(TokenKind::Newline);
         }
 
         match self.current_token_kind() {
-            TokenKind::Def => {
-                Stmt::FunctionDef(self.parse_function_definition(decorators, start_offset))
-            }
-            TokenKind::Class => {
-                Stmt::ClassDef(self.parse_class_definition(decorators, start_offset))
-            }
+            TokenKind::Def => Stmt::FunctionDef(self.parse_function_definition(decorators, start)),
+            TokenKind::Class => Stmt::ClassDef(self.parse_class_definition(decorators, start)),
             TokenKind::Async if self.peek() == TokenKind::Def => {
                 self.bump(TokenKind::Async);
 
+                // test_ok decorator_async_function
+                // @decorator
+                // async def foo(): ...
                 Stmt::FunctionDef(ast::StmtFunctionDef {
                     is_async: true,
-                    ..self.parse_function_definition(decorators, start_offset)
+                    ..self.parse_function_definition(decorators, start)
                 })
             }
             _ => {
+                // test_err decorator_unexpected_token
+                // @foo
+                // async with x: ...
+                // @foo
+                // x = 1
                 self.add_error(
                     ParseErrorType::OtherError(
-                        "expected class, function definition or async function definition after decorator".to_string(),
+                        "Expected class, function definition or async function definition after decorator".to_string(),
                     ),
                     self.current_token_range(),
                 );
+
+                // TODO(dhruvmanila): It seems that this recovery drops all the parsed
+                // decorators. Maybe we could convert them into statement expression
+                // with a flag indicating that this expression is part of a decorator.
+                // It's only possible to keep them if it's a function or class definition.
+                // We could possibly keep them if there's indentation error:
+                //
```

Comment:
Could we convert them into a binary expression where the left side is missing? TypeScript parses this as a `Decorator` and creates a `MissingDeclaration` node. https://astexplorer.net/#/gist/a5f4d9eabdade2111a1423af39012cb4/f58e76cd625d31670a149500f8cd3a23908a11dd

Reply:
Yeah, I've put it down as a possible solution. We would need to convert each decorator into a binary expression statement.